evidence/pool.go

```diff
@@ -109,6 +116,15 @@ func (evpool *Pool) Update(state sm.State) {
 	// update the state
 	evpool.updateState(state)
 
+	if len(evpool.consensusEvBuffer) != 0 {
+		// load evidence from buffer to clist for gossiping
+		for _, ev := range evpool.consensusEvBuffer {
+			evpool.evidenceList.PushBack((ev))
+		}
+		// flush evidence buffer
+		evpool.consensusEvBuffer = []types.Evidence{}
+	}
```

Comment:
We may want to check that the state height is sufficient for the evidence. I'm not familiar with this code, but there may be cases where evidence is produced and buffered, but the node restarts before the block is committed, and the node starts replaying blocks from an earlier height (e.g. if the app does not save state to disk for every height).

Reply:
Eh, since we buffer evidence in memory, it's going to be lost in the above scenario so it doesn't apply. A different scenario might be if the node receives any evidence while it's doing catchup with the chain (e.g. it may be lagging behind the head by 20 blocks).

Reply:
In the case when the evidence module restarts we already take everything from the evidence db and load it into in-memory list so it can be gossiped to peers.

Reply:
Yeah, that should be fine. The case I'd like to try to avoid is when `Update` is called before the height we're actually waiting for, for whatever reason - i.e. let's check that `state.LastBlockHeight` is `>=` the height of the evidence we'd like to submit.

Reply:
Interesting thing I just noticed is that before gossiping evidence we check that the peer we are sending evidence to is at a height that is equal to or greater than the evidence. So I'm thinking that this should stop us sending duplicate vote evidence prematurely
```
	var (
		peerHeight   = peerState.GetHeight()
		params       = evR.evpool.State().ConsensusParams.Evidence
		ageNumBlocks = peerHeight - evHeight
	)

	if peerHeight < evHeight { // peer is behind. sleep while he catches up
		return nil, true
	} el
```

Reply:
> let's check that state.LastBlockHeight is >= the height of the evidence 
Yeah I thought about that. I didn't think there would be an instance when it's not because Update panics if the next height is less than the last height.

Reply:
Good find. Yeah, that makes more sense - it doesn't matter if _we're_ caught up to the next block, it's the peer we're sending to that has to be caught up. Assuming `evHeight` is the height where the misbehavior occured, can we just change this predicate to `peerHeight <= evHeight`?

Reply:
If it's possible that we can jump back and update calls a previous height than perhaps we should change this panic :worried:

Reply:
I will change the predicate and see if this also solves it

Reply:
>If it's possible that we can jump back and update calls a previous height than perhaps we should change this panic 😟

Yeah, this can happen e.g. if the app doesn't save state and the node is restarted, where it will replay blocks from the height returned by the ABCI application. I'm not sure if this is actually a problem or not, since I don't know all of the details here (e.g. this code path may not trigger during block replay), but it's worth checking.

Reply:
I think during block replay it uses an empty evidence pool

Reply:
> I will change the predicate and see if this also solves it

This seems to work

Reply:
Awesome, that seems like a much cleaner solution.